src/lib/optimizer/strategy/join_to_predicate_rewrite_rule.cpp

```diff
@@ -0,0 +1,200 @@
+        return LQPVisitation::VisitInputs;
+      }
+      auto removable_side = join_node->get_unused_input();
+      if ((removable_side) || (join_node->join_mode == JoinMode::Semi)) {
+        std::shared_ptr<PredicateNode> valid_predicate = nullptr;
+        const auto can_rewrite = _check_rewrite_validity(join_node, removable_side, valid_predicate);
+        if (can_rewrite) {
+          rewritable_nodes.push_back(join_node);
+          removable_sides.push_back(*removable_side);
+          valid_predicates.push_back(valid_predicate);
+        }
+      }
+    }
+    return LQPVisitation::VisitInputs;
+  });
+
+  const auto n_rewritables = rewritable_nodes.size();
+  for (auto index = size_t{0}; index < n_rewritables; ++index) {
+    _perform_rewrite(rewritable_nodes[index], removable_sides[index], valid_predicates[index]);
+  }
+}
+
+bool JoinToPredicateRewriteRule::_check_rewrite_validity(const std::shared_ptr<JoinNode>& join_node,
+                                                         std::optional<LQPInputSide>& removable_side,
+                                                         std::shared_ptr<PredicateNode>& valid_predicate) const {
+  std::shared_ptr<AbstractLQPNode> removable_subtree = nullptr;
+
+  if (removable_side) {
+    removable_subtree = join_node->input(*removable_side);
+  } else {
+    // we know the join can only be a semi join in this case, where the right input is the one removed
+    auto r = LQPInputSide::Right;
+    removable_side = r;
+    removable_subtree = join_node->right_input();
+  }
+
+  std::shared_ptr<BinaryPredicateExpression> join_predicate = nullptr;
+  const auto& join_predicates = join_node->join_predicates();
+  for (auto& predicate : join_predicates) {
+    join_predicate = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate);
+    if (join_predicate) {
+      break;
+    }
+  }
+
+  DebugAssert(join_predicate, "A Join must have at least one BinaryPredicateExpression.");
+
+  std::shared_ptr<AbstractExpression> exchangable_column_expr = nullptr;
+
+  if (expression_evaluable_on_lqp(join_predicate->left_operand(), *removable_subtree)) {
+    exchangable_column_expr = join_predicate->left_operand();
+  } else if (expression_evaluable_on_lqp(join_predicate->right_operand(), *removable_subtree)) {
+    exchangable_column_expr = join_predicate->right_operand();
+  }
+
+  DebugAssert(exchangable_column_expr,
+              "Neither column of the join predicate could be evaluated on the removable input.");
+
+  // Check for uniqueness
+  auto testable_expressions = ExpressionUnorderedSet{};
+  testable_expressions.insert(exchangable_column_expr);
+
+  if (!removable_subtree->has_matching_unique_constraint(testable_expressions)) {
+    return false;
+  }
+
+  // Now, we look for a predicate that can be used inside the substituting table scan node.
+  visit_lqp(removable_subtree, [&removable_subtree, &valid_predicate](auto& current_node) {
```

Comment:
The visitation logic is quite complex. Please provide a comment that explains the algorithm and overall approach.